### ISSUE TITLE
General question: how solid is this tool?

### ISSUE THREAD

**Issue by zizipoil** [keplergl/kepler.gl#1425]
Hi there,

I have a general question over this tool/library: how solid/robust is it?

I stumble upon that tool some years ago and haven't had any real visualization project I could use it for before recently.
So these last days I happily started to try using it and convince my colleagues that we should integrate it in our portal.

But... here is a non-exhaustive list of some things that surprised me:

- broken trip visualization.
This used to be one of the [coolest visualizations](https://www.youtube.com/watch?v=OwwG865Qjh0).
Couldn't make it work. Couldn't find a simple example dataset to make it work.
Found a few [open issues](https://github.com/keplergl/kepler.gl/issues/397) referring to this problem.
Is it really broken? It used to work.

- very difficult to find the proper starting examples.
Depending if you come from kepler.gl, npm or vis.academy you are sent to different repos.
For reference, I wasn't even able to yarn install the 101 demo [here](https://github.com/uber-common/vis-academy/tree/master/src/demos/kepler.gl/0-starting-code) It breaks because of an archive behind a password protected repo.
I got better luck with the examples on this repository

- many open issues assigned to one person. Some opened for 2yr +
I don't know what to think about it.

So to summarize:
- I really like this tool and want to integrate it in my portal
- I see there's a very active community working on it
BUT
- I see regressions and broken features between 2 years ago and now
- I see a substantial amount of old issues still opened

Could someone shine some light on the situation?
Should I invest time in trying to integrate the tool or should I try using alternatives (deck.gl, echarts, mapbox...)?

**Comment by xsherlockpl** [comment]
Honestly, I see little user feedback  based fixes in the last 2 releases when I started using kepler.  Had to hire a top developer to fix the problems I had with it and add my features,  apart from it I find it quite reliable and the performance is top notch compared anything else I have tried.

**Comment by ibgreen** [comment]
> Had to hire a top developer to fix the problems I had with it and add my features

Did you then contribute those fixes and features back to kepler.gl? Like most open source project, kepler.gl is only the sum of its contributions. 

Over the last year, almost all contributions to kepler.gl come from the original creators working at a single company (Unfolded). Unfolded's focus over the last year has not been on adding features but on improving the internal structure of kepler.gl to support the products we are building on top of it. The work has included extensive investments in refactors, API improvements, typescript introduction etc that would all be valuable to anyone programming in kepler.gl and will unblock more advanced feature developments down the road.

So kepler.gl keeps getting (significantly) better, it just hasn't been adding many end-user features lately. 

Also keep in mind that as great as the kepler.gl web app is, ultimately it is really only a demo app for the power of the underlying kepler.gl framework which is where all the magic happens.

As a final comment, wearing both Unfolded and UCF hats, we would love to invite additional contributors to the project. Note that kepler.gl is under open governance in the Linux Foundation / UCF so it is very easy to collaborate, participate in planning and even get a say in decisions.

**Comment by ibgreen** [comment]
> very difficult to find the proper starting examples. Depending if you come from kepler.gl, npm or vis.academy you are sent to different repos.

Good feedback.
- Vis Academy was an independent effort and the contributors have unfortunately moved on to new jobs. We have been hoping someone would step in, and had a couple of indications of interest that did not pan out. Unless we can find someone to do a pass on them we may want to take down that site to avoid confusion.
- npm - Took a quick look but couldn't spot the bad link(s), perhaps you can just open issue with details (or even a PR)?

> many open issues assigned to one person. Some opened for 2yr +

Fair point, kepler needs an issue cleanup pass.

**Comment by zizipoil** [comment]
>     * Vis Academy was an independent effort and the contributors have unfortunately moved on to new jobs. We have been hoping someone would step in, and had a couple of indications of interest that did not pan out. Unless we can find someone to do a pass on them we may want to take down that site to avoid confusion.

Probably the best would be to have the vis academy website clearly state that it is outdated and send users to the correct place (this repo's example folder) for accurate information.

I found a few references to vis academy on that [page](https://docs.kepler.gl/).
I assume it's generated from this [readme](https://github.com/keplergl/kepler.gl/blob/70687cab9bf79af4eee3c750f2042d560519bcbe/typedoc/README.md)? If so I will submit a pull request when I find a bit of time.

>     * npm - Took a quick look but couldn't spot the bad link(s), perhaps you can just open issue with details (or even a PR)?

I think NPM lead me to realize this is the official kepler gl repo. No issue there.

> Fair point, kepler needs an issue cleanup pass.

To add on top of this, have a look at [this issue](https://github.com/keplergl/kepler.gl/issues/200): It's a typical example of a zombie user request revived after several years of inactivity.
The good part is seeing users trying to help each other. Pity no one steps in to propose a fix.

It seems that there are only [2 main contributors keeping the project alive](https://github.com/keplergl/kepler.gl/graphs/contributors)

My 2 cents:
- this project seems to have a large community
- as a new comer I find it very difficult to find accurate information
- as it is, I feel my lack of understanding of its design and architecture makes it very difficult to make any kind of meaningful contribution to it. 

Maybe I just missed the proper documentation?
For now the clearest 101 introduction I found was [this blog article from Zoba](https://medium.com/zoba-blog/integrating-with-kepler-gl-for-visual-based-spatial-analytics-9cfe4235423d)

**Comment by xsherlockpl** [comment]
> Did you then contribute those fixes and features back to kepler.gl? Like most open source project, kepler.gl is only the sum of its contributions.

Not yet, but we certainly should and will! While some of our mods are proprietary telecom drivetest stuff, some maybe could be useful for everybody like those extra stats on filters.

![image](https://user-images.githubusercontent.com/23699880/109150790-a21ccd00-7769-11eb-9dbf-58d031f0c0e9.png)

**Comment by zizipoil** [comment]
@ibgreen 

I've just had a look at your product Unfolded. Very nice!
It seems mainly built on top of Kepler.
Are you considering pushing some features back to Kepler too?

**Comment by ibgreen** [comment]
> I've just had a look at your product Unfolded. Looks very nice and robust. It seems mainly built on top of Kepler. 

Thanks! It is true that Unfolded Studio does use kepler.gl as an open core but we have a wide range of deep enhancements already in the product and in development. Also Unfolded Studio is just one of our products (we also have Data SDK, Map SDK, Data Catalog etc).  

As you know kepler.gl is front-end only, and as good as it is, there is so much more power we can add now that we also have a back-end, and we are iterating fast on that.

> I also see that @heshan0131 is part of the team.

Yes both Shan and @macrigiuseppe are part of Unfolded.

> Are you considering to push some features back to Kepler in the future?

Yes, we might potentially do that for features that do not require back-end support. At the moment we do not want it to be a one-way street though. We worked with Uber to move kepler.gl into the Linux foundation with the hope of getting interest and contributions from other companies and if that happens it would certainly increase our motivation.

I think an important take-away is that Uber has generously handed kepler.gl over to the community. And through Unfolded, there is still a company backing the project and all the key contributors are still around and involved with kepler.gl.

**Comment by ibgreen** [comment]
> Not yet, but we certainly should and will! some of our mods ... maybe could be useful for everybody like those extra stats on filters.

This looks like a great feature.

**Comment by ibgreen** [comment]
> as a new comer I find it very difficult to find accurate information

This is an important discussion. I think we should open a separate issue with actionable documentation improvement suggestions, as this issue per title is more of discussion around the health of the kepler.gl project and its community

There are pretty extensive docs https://docs.kepler.gl/

Perhaps in the separate issue you could provide a bulleted list of shortcomings in these docs?

> as it is, I feel my lack of understanding of its design and architecture makes it very difficult to make any kind of meaningful contribution to it.

Yes the kepler.gl architecture is not simple. And this one of the main reasons why almost all Unfolded contributions this year have been around refactors, improved typing etc. All investments in making kepler.gl more robust and easier to work with. Heavy work, but someone has to do it...

**Comment by heshan0131** [comment]
> > Did you then contribute those fixes and features back to kepler.gl? Like most open source project, kepler.gl is only the sum of its contributions.
> 
> Not yet, but we certainly should and will! While some of our mods are proprietary telecom drivetest stuff, some maybe could be useful for everybody like those extra stats on filters.
> 
> ![image](https://user-images.githubusercontent.com/23699880/109150790-a21ccd00-7769-11eb-9dbf-58d031f0c0e9.png)

This is a nice feature to add to the filter widget. For features like this, a good way to get our support is to create an issue, add the mockup and propose implementation details. We will be able to support you contribute it into kepler. There are some elements here e.g. min, max, avg are generalized enough to be added to kepler. 

In general. When we don't always have the resources to build every feature requested by the community. We'd love for people to propose solutions and contribute to the codebase. We have developed features in the past by providing full support to outside contributions. Examples would be the cloud provider with Carto, the localization with Gispo, geocoder and tooltip formatting with 7ParkData. And in all of those cases, our kepler.gl developers actually contributed the same amount if not more work alongside external contributors.

As Ib mentioned, the success of an OSS project is the sum of its contributors. We are a small team of only 3 developers actively maintaining kepler.gl. We want to be enablers more than sole contributors.
